src/java/org/apache/cassandra/index/sai/disk/vector/CompactionGraph.java

```diff
@@ -0,0 +1,347 @@
+                                        indexConfig.getConstructionBeamWidth(),
+                                        1.2f,
+                                        dimension > 3 ? 1.2f : 1.4f,
+                                        PhysicalCoreExecutor.pool(), ForkJoinPool.commonPool());
+
+        var indexFile = descriptor.fileFor(IndexComponent.TERMS_DATA, context);
+        termsOffset = (indexFile.exists() ? indexFile.length() : 0)
+                      + SAICodecUtils.headerSize();
+        writer = new OnDiskGraphIndexWriter.Builder(builder.getGraph(), indexFile.toPath())
+                 .withStartOffset(termsOffset)
+                 .with(new InlineVectors(dimension))
+                 .withMapper(new OnDiskGraphIndexWriter.IdentityMapper())
+                 .build();
+        writer.getOutput().seek(indexFile.length());
+        SAICodecUtils.writeHeader(SAICodecUtils.toLuceneOutput(writer.getOutput()));
+        inlineVectors = new InlineVectorValues(dimension, writer);
+        pqVectorsList = new ArrayList<>(entriesAllocated);
+        pqVectors = new PQVectors(compressor, pqVectorsList);
+        // VSTODO add LVQ
+        builder.setBuildScoreProvider(BuildScoreProvider.pqBuildScoreProvider(similarityFunction, inlineVectors, pqVectors));
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        // this gets called in finally{} blocks, so use closeQuietly to avoid generating additional exceptions
+        FileUtils.closeQuietly(inlineVectors);
+        FileUtils.closeQuietly(writer);
+        FileUtils.closeQuietly(postingsMap);
+    }
+
+    public int size()
+    {
+        return builder.getGraph().size();
+    }
+
+    public boolean isEmpty()
+    {
+        return postingsMap.values().stream().allMatch(VectorPostings::isEmpty);
+    }
+
+    /**
+     * @return the incremental bytes ysed by adding the given vector to the index
+     */
+    public AsyncAdd maybeAddVector(ByteBuffer term, Integer key) throws IOException
+    {
+        assert term != null && term.remaining() != 0;
+
+        var vector = vts.createFloatVector(serializer.deserializeFloatArray(term));
+        // Validate the vector.  Since we are compacting, invalid vectors are ignored instead of failing the operation.
+        try
+        {
+            VectorValidation.validateIndexable(vector, similarityFunction);
+        }
+        catch (InvalidRequestException e)
+        {
+            if (StorageService.instance.isInitialized())
+                logger.trace("Ignoring invalid vector during index build against existing data: {}", (Object) e);
+            else
+                logger.trace("Ignoring invalid vector during commitlog replay: {}", (Object) e);
+            return new AsyncAdd(0);
+        }
+
+        var bytesUsed = 0L;
+        var postings = postingsMap.get(vector);
+        if (postings == null)
+        {
+            if (!postingsMap.containsKey(vector))
+            {
```

Comment:
Can we simplify this now that `maybeAddVector` is called synchronously? I guess we need to be careful because we are operating in the `addInternalAsync` method, but it seems like there is some complexity we don't need in this version.